src/main/scala/info/bethard/timenorm/formal/Types.scala

```diff
@@ -294,12 +307,129 @@ case class IntervalSubIntervalIntersection(interval: Interval, subInterval: Repe
   val end = ???
 }
 
-trait RepeatingInterval extends TimeExpression
-case class UnitRepeatingInterval(unit: TemporalUnit, modifier: Modifier) extends RepeatingInterval
-case class FieldRepeatingInterval(field: TemporalField, value: Long, modifier: Modifier) extends RepeatingInterval
-case class NumberedRepeatingInterval(repeatingInterval: RepeatingInterval, number: Number) extends RepeatingInterval
-case class RepeatingIntervalUnion(repeatingIntervals: Set[RepeatingInterval]) extends RepeatingInterval
-case class RepeatingIntervalIntersection(repeatingIntervals: Set[RepeatingInterval]) extends RepeatingInterval
-case class IntervalAsRepeatingInterval(interval: Interval) extends RepeatingInterval
+trait RepeatingInterval extends TimeExpression {
+  def preceding(interval: Interval): Iterator[Interval]
+  def following(interval: Interval): Iterator[Interval]
+
+  private[formal] def truncate(ldt: LocalDateTime, tUnit: TemporalUnit): LocalDateTime = {
+    tUnit match {
+      case ChronoUnit.CENTURIES => LocalDateTime.of(ldt.getYear/100*100,1,1,0,0)
+      case ChronoUnit.DECADES => LocalDateTime.of(ldt.getYear/10*10,1,1,0,0)
+      case ChronoUnit.YEARS => ldt.withDayOfYear(1).truncatedTo(ChronoUnit.DAYS)
+      case ChronoUnit.MONTHS => ldt.withDayOfMonth(1).truncatedTo(ChronoUnit.DAYS)
+      case ChronoUnit.WEEKS =>
+        ldt.withDayOfYear(ldt.getDayOfYear-ldt.getDayOfWeek.getValue).truncatedTo(ChronoUnit.DAYS)
+      case _ => ldt.truncatedTo(tUnit)
+    }
+  }
+}
+
+case class UnitRepeatingInterval(unit: TemporalUnit, modifier: Modifier) extends RepeatingInterval {
+  override def preceding(interval: Interval): Iterator[Interval] = {
+    var end = truncate( interval.start, unit ).plus(1,unit)
+    var start = end.minus(1, unit)
+
+    Iterator.continually {
+      start = start.minus(1, unit)
+      end = end.minus(1, unit)
+      SimpleInterval(start, end)
+    }
+  }
+
+  override def following(interval: Interval): Iterator[Interval] = {
+    var start = truncate(interval.end, unit)
+    var end = start.plus(1,unit)
+
+    Iterator.continually {
+      start = start.plus(1, unit)
+      end = end.plus(1, unit)
+      SimpleInterval(start, end)
+    }
+  }
+}
+
+
+case class FieldRepeatingInterval(field: TemporalField, value: Long, modifier: Modifier) extends RepeatingInterval {
+  if ( !field.range.isValidValue(value))
+    throw new IllegalArgumentException
+
+  override def preceding(interval: Interval): Iterator[Interval] = {
+    var start = truncate(interval.start.`with`(field,value), field.getBaseUnit)
+
+    if (start.isBefore(interval.start))
+      start = start.plus(1,field.getRangeUnit)
+
+    var end = start.plus(1,field.getBaseUnit)
```

Comment:
That was added because we discussed protecting FieldRepeatingIntervals from long values that are not valid for the TemporalField range at the time of construction. However, a DateTimeException will be thrown when the Iterator is created so I've changed the test to reflect that.